### DISCUSSION TITLE
Any way to $watch if an array ?

### DISCUSSION THREAD

**Discussion by yabdab** [alpinejs/alpine#3922] [1. Help]
Is there a way to $watch a variable  if the value is an array or object ( containing an array ).

I don't need to know "what" value changed, but that the arrays content changed in some way.

This is what I have , but changing values in the array value does not trigger console to log anything?

```
x-init="$watch('itemsArray', value => console.log(value));"
```

Is this even possible?

**Comment by ekwoka** [comment]
This should totally work.

Do you have a reproduction?

**Comment by yabdab** [comment]
It sort of works, but I get this error every time a change is made to itemsArray?

```
alpine.js:2 Uncaught TypeError: Converting circular structure to JSON
    --> starting at object with constructor 'Object'
    |     index 0 -> object with constructor 'Object'
    --- property 'itemsArray' closes the circle
    at JSON.stringify (<anonymous>)
    at alpine.js:2:28546
    at fe (alpine.js:2:6928)
    at alpine.js:2:6554
    at ne (alpine.js:2:5343)
    at alpine.js:2:28534
    at Array.n (alpine.js:2:42436)
    at l (alpine.js:2:166)
```

**Reply by yabdab** [reply]
I am pushing the component $data into the itemsArray as they load. 
Not sure what I need to do to fix this "circular structure"?
Guess I will study up on that. 😖

**Reply by SimoTod** [reply]
You get that error if the element you push contains the array itself. In order to make $watch work at any level, the array must be serializable

**Reply by yabdab** [reply]
@SimoTod Yea, strangely enough I just figured it out from one of your previous posts. 
I was pushing $data into the array and it included the parent. So I only needed the "closest" data.

I used your code `itemsArray.push(Alpine.closestDataStack($el)[0]);` to get only what I needed. 

This fixed everything 🎉

**Reply by bennadel** [reply]
I just ran into this as well. Instead of pushing the `$data`, though, I was pushing `element`. This also causes a circular reference because the `element`'s `_x_dataStack` contains a reference to the parent `$data` proxy, which in turn contains an array of `element` entries, each of which contain an `_x_dataStack` that points back to the parent (and so on).

It seems to become problematic because of this line internally to `watch()`:

```js
 // JSON.stringify touches every single property at any level enabling deep watching
JSON.stringify(value)
```

Anyway, I'm not looking for an answer here, I'm just updating the comment thread for anyone else who comes after me. I don't believe that what I am trying to do is possible given the current mechanics.

**Reply by ekwoka** [reply]
Yes, watch only works because it stringifies the value to link into the reactivity of every value on the tree.

This could be avoided with a custom walker that accesses everything, but really, this is an XY problem. If you're somehow passing the context into itself....that's likely just a poor way to do that thing.

It would definitely require more context to help rectify.

**Reply by bennadel** [reply]
@ekwoka all good - I'm just doing R&D, seeing what's possible.
